Recaptcha

benbieler opened the pull request:
Resolves #333 
- [x] Frontend using react-recaptcha.
- [x] Built constraint for DTO.
- [x] Unit test validator

Ma27 commented on app/Resources/views/index.html.twig:
```diff
@@ -20,6 +20,7 @@
     </head>
     <body>
         <div id="content"></div>
+        <script src="https://www.google.com/recaptcha/api.js?onload=onloadCallback&render=explicit" async defer></script>
```
dowe need to use a CDN or is it pissible to resolve this as NPM dependenxy?

Ma27 replied:
unless just ignore that😉

Ma27 commented on src/AppBundle/Model/User/DTO/CreateUserDTO.php:
```diff
@@ -15,6 +15,7 @@
 namespace AppBundle\Model\User\DTO;
 
 use AppBundle\Model\User\User;
+use AppBundle\Validator\Constraints as Captcha;
```
please import the class as-is and not a prefix (look at Locale and UniqueProperty)

Ma27 commented on src/AppBundle/Model/User/Handler/CreateUserHandler.php:
```diff
@@ -101,7 +104,12 @@ public function __invoke(CreateUserDTO $userDTO)
+            ];
+
+            $isUnique       = count($this->validator->validate($activationKey, new UniqueProperty($options))) === 0;
+            $IsCaptchaValid = count($this->validator->validate($activationKey, new Recaptcha($hash))) === 0;
```
can be reverted, during the evaluation with the command bus a middleware validates the DTO, this is only for the approval key generation.

Ma27 commented on src/AppBundle/Tests/Unit/Validator/Constraints/RecaptchaValidatorTest.php:
```diff
@@ -0,0 +1,17 @@
+
+namespace AppBundle\Tests\Unit\Validator\Constraints;
+
+class RecaptchaValidator
```
test files need a test postfix

Ma27 commented on src/AppBundle/Tests/Unit/Validator/Constraints/RecaptchaValidatorTest.php:
```diff
@@ -0,0 +1,17 @@
+<?php
+/*
```
every php file should have a strict type declaration

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,60 @@
+     * Validation message, to check if captcha is correct.
+     *
+     * @param mixed      $value
+     * @param Constraint $constraint
```
instead of copying the whole docs, just put {@inheritdoc} into the block and that's it

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,60 @@
+
+        $recaptcha = new \ReCaptcha\ReCaptcha($this->container->getParameter('secret_key'));
+
+        $check = $recaptcha->verify($constraint->recaptchaHash, 'http://sententiaregum.dev/');
```
the url sententiaregum.dev should be stored as parameter, too.

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,60 @@
+            throw new UnexpectedTypeException($constraint, Recaptcha::class);
+        }
+
+        $recaptcha = new \ReCaptcha\ReCaptcha($this->container->getParameter('secret_key'));
```
pls import the class as use statement

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @param Container $container
+     */
+    public function __construct(Container $container)
```
the parameters can be injected directly (using 'arguments: ["%param%"]' in the service definition. Much simpler than injecting the container

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,60 @@
+     */
+    public function __construct(Container $container)
+    {
+        $this->container = $container;
```
it's a best practice here to define every property in PHP in class scope before using them in methods (private $dependency at the top for instance)

Ma27 commented on src/AppBundle/Validator/Constraints/Recaptcha.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Hash which the user creates.
+     *
+     * @var RecaptchaHash (required!!)
```
not necessary to add here a note that it's required. If it doesn't exist, an exception will be thrown

Ma27 commented on src/AppBundle/Validator/Constraints/Recaptcha.php:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @var string
+     */
+    public $message = 'You are not allowed to access this API endpoint!';
```
correct me if I'm wrong, but should $message contain the translation key? (Captch_Invalid?)

Ma27 replied:
would rather use as default something like "Invalid captcha". The translation is added in the DTO.

Ma27 commented:
for unittesting: how about not instantiating the recaptcha class, but injecting it in the constructor? you could simply define it in the services.yml in the resources dir and inject a phpunit mock in the test.

Ma27 commented on docs/setup/custom_config.md:
```diff
@@ -3,5 +3,8 @@
 If you'd like to override configuration values, you can create for every environment in the *app/config* directory a custom configuration file:
 
     app/config/local_config_{environment}.yml
+If you want to create a new account, you will have to add a line called "secret_key" to your parameters.yml.
```
add the value with default to '~' in the parameters.yml.dist then composer will add it automatically to the parameters file. It's only necessary to customize the value. (sry if I was unclear about this when chatting)

Ma27 replied:
would be great if you could update the docs for that

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,60 @@
+
+        if (!$check->isSuccess()) {
+            $this->context->buildViolation($constraint->message)
+                ->addViolation();
```
pls add invalidvalue and atpath as it might help with debugging in the future (have a look at the UniquePropertyValidator)

Ma27 commented on src/AppBundle/Resources/config/validators.yml:
```diff
@@ -11,3 +11,7 @@ services:
       - "@doctrine"
     tags:
       - { name: validator.constraint_validator, alias: "app.validator.unique_property" }
+  app.validator.recaptchaKey:
```
pls use underscore rather than camel case for service ids

Ma27 commented on src/Frontend/components/portal/signup/Form.js:
```diff
@@ -107,12 +109,26 @@ export default class Form extends Component {
           options={{ de: 'Deutsch (Deutschland)', en: 'English (USA)' }} />
-        <LoadableButtonBar btnLabel={this.helper.getFormFieldAlias('button')} progress={this.state.progress} />
+          <Recaptcha
+              sitekey="6Lds4igTAAAAAJDx1KvNfznuELzNauwKu21ODAWR"
```
to hide such keys completely before merging, I recommend to build a config file (e.g. Frontend/config/recaptcha.js) ignored by git which exports a string, nothing more. It might be generated by NPM postinstall if it doesn't exist and to test registration the key must be added there, too (this note that the key must be added also there should be added to the docs)

Ma27 commented on package.json:
```diff
@@ -10,7 +10,8 @@
     "test": "mocha src/Frontend/test --recursive --ui bdd --compilers js:babel-core/register --reporter dot --require babel-polyfill --require src/Frontend/test/DOM.js",
-    "watch": "webpack --watch"
+    "watch": "webpack --watch",
+    "postinstall": "touch ./src/Frontend/config/recaptcha.js"
```
it shouldn't be created if it exists already and to the new file an 'export default ""' should be added. Would recommend to move it into a custom shell in bin/

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,61 @@
+
+        $recaptcha = new ReCaptcha($this->container->getParameter('secret_key'));
+
+        $check = $recaptcha->verify($constraint->recaptchaHash, $this->container->getParameter('site_url'));
```
actually $value is the one to be validated, right? $constraint->recaptchaHash can be removed right?

Ma27 commented on src/AppBundle/Model/User/Handler/CreateUserHandler.php:
```diff
@@ -85,8 +85,10 @@ public function __invoke(CreateUserDTO $userDTO)
-        $isUnique      = false;
+        $rounds         = 0;
+        $isUnique       = false;
+        $IsCaptchaValid = false;
```
this one is not needed anymore right?

Ma27 commented:
would be great if you could also update the ApiDoc annotation in the UserController (the action for user creation has this annotation to document our REST API)

Ma27 commented on src/AppBundle/Model/User/Handler/CreateUserHandler.php:
```diff
@@ -101,7 +102,7 @@ public function __invoke(CreateUserDTO $userDTO)
             ];
 
-            $isUnique = count($this->validator->validate($activationKey, new UniqueProperty($options))) === 0;
+            $isUnique       = count($this->validator->validate($activationKey, new UniqueProperty($options))) === 0;
```
here and above the alignment before the assignment statement is invalid

Ma27 commented on src/AppBundle/Tests/Unit/Validator/Constraints/RecaptchaValidatorTest.php:
```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
```
to keep it consistent with all other files the structure should look like this:

``` php
<?php

/*
 * top docblock
 */

declare(strict_types=1);

namespace Namespace;

class Class
{
}
```

(sry if my previous comment here was misleading)

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -0,0 +1,61 @@
+        if (!$check->isSuccess()) {
+            $this->context->buildViolation($constraint->message)
+                ->setInvalidValue($value)
+                ->atPath($constraint->propertyPath)
```
thought about it again, remove `atPath` and keep `setInvalidValue`. My previous comment that `atPath` is necessary here too was garbage, sry.

Ma27 commented on src/Frontend/components/portal/signup/Form.js:
```diff
@@ -107,12 +110,33 @@ export default class Form extends Component {
+              onloadCallback={callback}
+              verifyCallback={this.verifyCallback.bind(this)}
+          />
+          <LoadableButtonBar btnLabel={this.helper.getFormFieldAlias('button')} progress={this.state.progress} />
```
two spaces indent here, right?

benbieler replied:
Yea, sry. I should really configure my IDE correctly

Ma27 replied:
well, actually the react eslint plugin isn't configured properly, need to fix this when I've got time to work on Sententiaregum again :)

Ma27 commented on src/Frontend/components/portal/signup/Form.js:
```diff
@@ -107,12 +110,33 @@ export default class Form extends Component {
+   * @returns {void}
+   */
+  verifyCallback(response) {
+    var newState = update(this.state, {
```
nice, didn't know this addon until now, thanks :+1: 
need to do this rather than the deep-assign stuff I've built: https://github.com/Sententiaregum/Sententiaregum/issues/344

benbieler replied:
Jup. Sounds good

Ma27 commented on src/Frontend/components/portal/signup/Form.js:
```diff
@@ -71,13 +74,13 @@ export default class Form extends Component {
    * @returns {React.Element} The vDOM markup.
    */
   render() {
+    const callback = function () {};
```
write sth. like this:

``` javascript
const callback = () => {
};
```

Oneline function declarations look awkward to me.

Ma27 commented on package.json:
```diff
@@ -81,5 +83,8 @@
     "node": ">=6.0.0"
+  },
+  "bin": {
+    "postinstall": "touch ./src/Frontend/config/recaptcha.js"
```
not sure if you got what I wanted to say in my previous comment:

a postinstall script is fine, but the logic in there could be moved into a shell file (e.g. `generate-frontend-recaptcha.sh` placed in the `bin/` dir in the project's root).
Furthermore this file should check if a config file already exists before creating one (`if [ -e 'path/to/config' ];`).

And if a new one is created, something like this could be put into the file from shell: `export default ''` (the user should add the key here too if the registration is needed).

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -27,14 +26,27 @@
      *
-     * @param Container $container
+     * @param $secret_key
+     * @param $site_url
```
`@param string ${name}`

Ma27 commented on src/AppBundle/Validator/Constraints/RecaptchaValidator.php:
```diff
@@ -27,14 +26,27 @@
+     * @param $site_url
      */
-    public function __construct(Container $container)
+    public function __construct($secret_key, $site_url)
```
we do PHP7, so write `string $secretKey`.
and the params should be camelCase ($secretKey)

Ma27 commented on src/AppBundle/Model/User/Handler/CreateUserHandler.php:
```diff
@@ -85,8 +85,9 @@ public function __invoke(CreateUserDTO $userDTO)
-        $rounds        = 0;
-        $isUnique      = false;
+        $rounds         = 0;
+        $isUnique       = false;
```
I think this change should be reverted, right?

Ma27 commented on src/AppBundle/Validator/Constraints/Recaptcha.php:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @var string
+     */
+    public $recaptchaHash;
```
this one is not needed anymore, right?

Ma27 commented on src/AppBundle/Validator/Constraints/Recaptcha.php:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @var string
+     */
+    public $propertyPath;
```
this one is not needed anymore, right?

Ma27 commented on bin/generate-frontend-recaptcha.sh:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+if [ -e "./src/Frontend/config/recaptcha.js" ]; then
+    echo -e "\e[31m You already have a recaptcha.js file! \e[0m"
```
:+1: nice!

Ma27 commented:
seems as we've missed something. Will take ownership of this branch now and merge everything @benbieler

Ma27 committed:
Enhancement #333 #341 finalize PR

- exclude fixture data
- resolve race-condition caused by deference of the reCAPTCHA library
- fixed remaining tests

benbieler committed:
Enhancement #333 implement a captcha system to protect the registration with reCAPTCHA and a lightweight constraint validator

benbieler committed:
Enhancement #333 connect reCAPTCHA with the view

Ma27 commented:
ok, merged everything to master. Nice idea and work @benbieler :+1: !
